File: 142_Linked_List_Cycle_II/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        # Boundary Check
        if not head:
            return None
        
        # First Check Whether it is cirle LinkedList
        node_cycle = self.checkCycle(head)
        if not node_cycle:
            return None
        
        # Get the num of Cycle Nodes
        num_cycle = 1
        node_next = node_cycle.next
        while node_next != node_cycle:
            node_next = node_next.next
            num_cycle += 1
            
        # Use two pointer
        fast = head
        slow = head
        
        for _ in range(num_cycle):
            fast = fast.next
            
        while fast != slow:
             fast = fast.next
             slow = slow.next
        
        return slow
        
        
    def checkCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        # Boundary Check
        if not head:
            return None
        
        # Iteratively
        fast = head
        slow = head
        
        while fast:
            fast = fast.next
            
            if not fast:
                return None
            
            fast = fast.next
            slow = slow.next
            
            if fast == slow:
                return slow
        
        return None
            
        

File: 142_Linked_List_Cycle_II/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func detectCycle(head *ListNode) *ListNode {
    linkLen := lenCycle(head)
    if linkLen == -1 {
        return nil
    }
    
    slow := head
    fast := head
    
    for i := 0; i < linkLen; i++ {
        fast = fast.Next
    }
    
    for fast != slow {
        fast = fast.Next
        slow = slow.Next
    }
    
    return slow
}

func lenCycle(head *ListNode) int {
    if head == nil {
        return -1
    }
    
    fast := head.Next
    slow := head
    
    for fast != slow {
        if fast == nil || fast.Next == nil {
            return -1
        }
        fast = fast.Next.Next
        slow = slow.Next
    }
    
    linkLen := 1
    slow = slow.Next
    
    for slow != fast {
        slow = slow.Next
        linkLen ++
    }
    
    return linkLen
}
